libzecale/circuits/pairing/bls12_377_pairing.hpp

```diff
@@ -0,0 +1,332 @@
+    typedef libff::Fqe<libsnark::other_curve<ppT>> FqeT;
+
+    bls12_377_G2_proj<ppT> in_R;
+    bls12_377_G2_proj<ppT> out_R;
+
+    // TODO: Many of these intermediate Fqe_variables are only for clarity and
+    // replicate the references held by other gadgets (e.g. `A` refers to the
+    // same variable as `check_A.result`. Do an optimization pass and remove
+    // some of the redundancy.
+
+    // A = R.X * R.Y / 2
+    libsnark::Fqe_variable<ppT> A;
+    libsnark::Fqe_mul_gadget<ppT> check_A; // R.X * R.Y = 2_times_A
+
+    // B = R.Y^2
+    libsnark::Fqe_variable<ppT> B;
+    libsnark::Fqe_sqr_gadget<ppT> check_B; // R.Y^2 == B
+
+    // C = R.Z^2
+    libsnark::Fqe_variable<ppT> C;
+    libsnark::Fqe_sqr_gadget<ppT> check_C; // R.Z^2 == C
+
+    // D = 3 * C
+    // libsnark::Fqe_variable<ppT> D;
+
+    // E = b' * D
+    libsnark::Fqe_variable<ppT> E;
+
+    // F = 3 * E
+    libsnark::Fqe_variable<ppT> F;
+
+    // G = (B + F) / 2
+    libsnark::Fqe_variable<ppT> G;
+    // libsnark::Fqe_mul_by_lc_gadget<ppT> check_G; // 2 * G == B + F
+
+    // H = (Y + 2) ^ 2 - (B + C)
+    libsnark::Fqe_variable<ppT> H;
+    libsnark::Fqe_variable<ppT> Y_plus_Z;
+    libsnark::Fqe_variable<ppT> H_plus_B_plus_C;
+    libsnark::Fqe_sqr_gadget<ppT> check_H; // Y_plus_Z^2 == H + B + C
+
+    // I = E - B
+    libsnark::Fqe_variable<ppT> I;
+
+    // J = R.X^2
+    libsnark::Fqe_variable<ppT> J;
+    libsnark::Fqe_sqr_gadget<ppT> check_J; // R.X^2 == J
+
+    // E^2
+    libsnark::Fqe_variable<ppT> E_squared;
+    libsnark::Fqe_sqr_gadget<ppT> check_E_squared;
+
+    // G^2
+    libsnark::Fqe_variable<ppT> G_squared;
+    libsnark::Fqe_sqr_gadget<ppT> check_G_squared;
+
+    // B - F
+    libsnark::Fqe_variable<ppT> B_minus_F;
+
+    // out_R.X = A * (B - F)
+    libsnark::Fqe_mul_gadget<ppT> check_out_Rx;
+
+    // out_R.Y = G^2 - 3 * E^2
+    // check: 1 * G_squared_minus_3_E_squared == outRy
+    libsnark::Fqe_variable<ppT> G_squared_minus_3_E_squared;
+    libsnark::Fqe_mul_by_lc_gadget<ppT> check_out_Ry;
+
+    // out_R.Z = B * H
+    libsnark::Fqe_mul_gadget<ppT> check_out_Rz;
+
+    // ell_0 = xi * I
+    // ell_vw = -H
+    // ell_vv = 3 * J
+    bls12_377_ate_ell_coeffs<ppT> out_coeffs;
+
+    bls12_377_ate_dbl_gadget(
+        libsnark::protoboard<FqT> &pb,
+        const bls12_377_G2_proj<ppT> &R,
+        const std::string &annotation_prefix);
+
+    void generate_r1cs_constraints();
+
+    // R should already be assigned. Computes all internal values and
+    // outResult.
+    void generate_r1cs_witness(const libff::Fr<ppT> &two_inv);
+};
+
+template<typename ppT>
+class bls12_377_ate_add_gadget : public libsnark::gadget<libff::Fr<ppT>>
+{
+public:
+    typedef libff::Fq<libsnark::other_curve<ppT>> FqT;
+    typedef libff::Fqe<libsnark::other_curve<ppT>> FqeT;
+
+    libsnark::Fqe_variable<ppT> Q_X;
+    libsnark::Fqe_variable<ppT> Q_Y;
+    bls12_377_G2_proj<ppT> in_R;
+
+    // A = Qy * R.Z;
```

Comment:
Shouldn't it be `(Q_X, Q_Y)` instead of `(Qx, Qy)`?

Reply:
Changed